subsys/fs/nvs/nvs.c

```diff
@@ -128,13 +135,32 @@ static int nvs_flash_al_wrt(struct nvs_fs *fs, uint32_t addr, const void *data,
 		offset += blen;
 		data8 += blen;
 	}
-	if (len) {
-		memcpy(buf, data8, len);
-		(void)memset(buf + len, fs->flash_parameters->erase_value,
-			fs->flash_parameters->write_block_size - len);
 
-		rc = flash_write(fs->flash_device, offset, buf,
-				 fs->flash_parameters->write_block_size);
+	/* Write the remaining non-aligned data, if any, filling the unaligned bytes
+	 * with the erase value
+	 */
+	if (len || is_crc_present) {
+		pbuf = buf;
+		if (len) {
+			memcpy(pbuf, data8, len);
+			pbuf += len;
+		}
+		if (is_crc_present) {
+			memcpy(pbuf, &data_crc, sizeof(data_crc));
+			pbuf += sizeof(data_crc);
+			len += sizeof(data_crc);
+		}
+
+		/* Fill the remaining space (if any) in the last write unit */
+		unaligned_bytes_count = len % fs->flash_parameters->write_block_size;
+		(void)memset(pbuf, fs->flash_parameters->erase_value,
+			     fs->flash_parameters->write_block_size - unaligned_bytes_count);
+
+		/* Write the last data part (if any), followed by the CRC (if any), followed by
+		 * the padding needed to respect the write block size (if any)
+		 */
+		len += unaligned_bytes_count;
```

Comment:
I think this computation is wrong and should be :
len += fs->flash_parameters->write_block_size - unaligned_bytes_count;
Example :
write_block_size = 8
len = 7
len += sizeof(data_crc); => len = 11
unaligned_bytes_count = 3
len + unaligned_bytes_count; => len = 14

writes should be always a multiple of write_block_size
len + write_block_size - unaligned_bytes_count = 16

Reply:
Correct, the unaligned bytes count is `fs->flash_parameters->write_block_size - (len % fs->flash_parameters->write_block_size)`, good catch !